src/algorithm/bluestein.rs

```diff
@@ -0,0 +1,207 @@
+    inverse: bool,
+) {
+    let mut scratch = vec![Complex::zero(); fft.len()];
+    let scale = T::one() / T::from_usize(fft.len()).unwrap();
+    for (i, tw) in scratch.iter_mut().enumerate() {
+        if let Some(index) = {
+            if i < len {
+                Some((i as f64).powi(2))
+            } else if i > fft.len() - len {
+                Some(((i as f64) - (fft.len() as f64)).powi(2))
+            } else {
+                None
+            }
+        } {
+            let twiddle = calculate_twiddle(index, len)*scale;
+            if inverse {
+                *tw = twiddle;
+            }
+            else {
+                *tw = twiddle.conj();
+            }
+        } 
+    }
+    fft.process(&mut scratch, &mut twiddles[..]);
+}
+
+
+/// Calculate the "x" twiddles.
+fn calculate_x_twiddles<T: FFTnum>(
+    len: usize,
+    twiddles: &mut [Complex<T>],
+    inverse: bool,
+) {
+    if inverse {
+        for (i, tw) in twiddles.iter_mut().enumerate() {
+            *tw = calculate_twiddle(-(i as f64).powi(2), len);
+        }
+    }
+    else {
+        for (i, tw) in twiddles.iter_mut().enumerate() {
+            *tw = calculate_twiddle(-(i as f64).powi(2), len).conj();
+        }
+    }
+}
+
+
+impl<T: FFTnum > Bluesteins<T> {
+    /// Creates a FFT instance which will process inputs/outputs of size `len`. `inner_fft_fw.len()` and `inner_fft_inv.len()` must be the nearest 
+    /// power of two that is equal to or larger than `2 * len - 1`.
+    ///
+    /// Note that this constructor is quite expensive to run; This algorithm must run a FFT within the
+    /// constructor. 
+    pub fn new(len: usize, inner_fft_fw: Arc<FFT<T>>, inner_fft_inv: Arc<FFT<T>>, inverse: bool) -> Self {
+        let inner_fft_len = (2 * len - 1).checked_next_power_of_two().unwrap();
+        assert_eq!(inner_fft_len, inner_fft_fw.len(), "For Bluesteins algorithm, inner_fft_fw.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_fw.len());
+        assert_eq!(inner_fft_len, inner_fft_inv.len(), "For Bluesteins algorithm, inner_fft_inv.len() must be a power of to larger than or equal to 2*self.len() - 1. Expected {}, got {}", inner_fft_len, inner_fft_inv.len());
+
+        let mut w_twiddles = vec![Complex::zero(); inner_fft_len];
+        let mut x_twiddles = vec![Complex::zero(); len];
+        calculate_w_twiddles(len, &inner_fft_fw, &mut w_twiddles, inverse);
+        calculate_x_twiddles(len, &mut x_twiddles, inverse);
+        Self {
+            len,
+            inner_fft_fw,
+            inner_fft_inv,
+            w_twiddles: w_twiddles.into_boxed_slice(),
+            x_twiddles: x_twiddles.into_boxed_slice(),
+            inverse,
+        }
+    }
+
+    fn perform_fft(&self, input: &mut [Complex<T>], output: &mut [Complex<T>]) {
+        assert_eq!(self.len(), input.len());
+
+        let mut scratch_a = vec![Complex::zero(); self.inner_fft_fw.len()];
```

Comment:
Minor optimization: If we have to allocate in here, it might be wise to only do one allocation instead of two. IE, create a vec of twice the size, and use split_at_mut to get two slices.

Benefits would include less memory fragmentation, fewer allocator calls, improved memory locality. I don't know if it would actually be measurably faster though.

Reply:
Yes unfortunately it's not possible to avoid the allocation without changing the FFT trait so process() takes a &mut self.
But this is a good idea. I'm guessing that with overhead, the time to allocate a twice as long vector takes much less than twice the time.